h2o-core/src/test/java/water/UnknownHeartbeatTest.java

```diff
@@ -0,0 +1,40 @@
+package water;
+
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.nio.channels.DatagramChannel;
+
+import static org.junit.Assert.assertEquals;
+
+public class UnknownHeartbeatTest extends TestUtil{
+  @BeforeClass() public static void setup() {
+    stall_till_cloudsize(1);
+  }
+
+  @Test
+  public void testIgnoreUnknownHeartBeat() throws IOException, NoSuchMethodException, InvocationTargetException, IllegalAccessException {
+    HeartBeat hb = new HeartBeat();
+    hb._cloud_name_hash = 777;
+    hb._client = true;
+    hb._jar_md5 = H2O.SELF._heartbeat._jar_md5;
+
+    AutoBuffer ab = new AutoBuffer(H2O.SELF,H2O.MAX_PRIORITY);
+
+    Method putSpMethod = AutoBuffer.class.getDeclaredMethod("putSp", int.class);
```

Comment:
Why do you need to access the private method instead of calling one of the public put methods?

Reply:
`putSp` method is package private method for `AutoBuffer`. I wanted to mimic the behaviour of `putUdp` method, but I can't use that method right away since I need to change the port

Reply:
You can accomplish the exact same thing if you use public methods

put1(byte)
put2(char)

This will have the same effect.

You can also introduce a new method putUdp and have it parametrized (instead of using H2O.SELF) - this option is even better because the underlying implementation can change and your test can stop working without anyone noticing.

Reply:
Good point with the public methods. I changed that in the code. However I don't understand the second comment. I need to use H2O.SELF as H2O.SELF is the recipient of the fake heartbeat

Reply:
Okay, maybe I don't understand why can you not use putUdp method. Can you please elaborate?

Reply:
I can't use `putUdp` method as that method hardcodes the sender port to `H2O.SELF` port.
` _bb.putChar((char)H2O.H2O_PORT  ); // Outgoing port is always the sender's (me) port`

However in this test, we need to simulate request from different node, therefore I need to change the port number in here. I think it doesn't make sense to create a new parametrised `putUdp` method as it's use-case would be just this test. In all other use-cases we want this method to mark the current node as the sender

Reply:
I put comment into the test code to make it more clear

Reply:
_However in this test, we need to simulate request from different node, therefore I need to change the port number in here. I think it doesn't make sense to create a new parametrised putUdp method as it's use-case would be just this test. In all other use-cases we want this method to mark the current node as the sender_

This is exactly what I was thinking except for that I think it is a common practice :) When it is hard to test particular function you can create a more general version of the function where  certain things are parametrized. You delegated the original method to the more general one with the original static parameter. This is usually done with any code that deals with time or with "now".

I would do it this way, it is much cleaner approach.

Reply:
oki, done :)